daemon/monitor.go

```diff
@@ -189,9 +191,11 @@ func (daemon *Daemon) ProcessEvent(id string, e libcontainerdtypes.EventType, ei
 			execConfig.StreamConfig.Wait(ctx)
 			cancel()
 
-			if err := execConfig.CloseStreams(); err != nil {
+			ctx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
+			if err := execConfig.CloseStreams(ctx); err != nil {
```

Comment:
this change doesn't fix the issue afaict because the blocking call to `CloseStreams` doesn't come from here, but earlier from `handleContainerExit` which calls `tsk.Delete(ctx)` which calls `io.Close()`, which doesn't take a context/will still block forever.

Reply:
Yes, you're right. The code has changed since this PR was created and now it hangs earlier

Reply:
Btw, I did a bit more analysis on why the code was not getting stuck on that `Delete` earlier, but is getting stuck there now.

In past this `handleContainerExit` was calling [`Process.Delete`](https://github.com/containerd/containerd/blob/157fe6a/process.go#L204), not [`Task.Delete`](https://github.com/containerd/containerd/blob/2aeac9a/task.go#L299)

(to be more exact, call chain was [`handleContainerExit`](https://github.com/moby/moby/blob/b5f863c67e6ffbaadedb94c885c3c50c625e4eb8/daemon/monitor.go#L30) -> [`client.DeleteTask`](https://github.com/moby/moby/blob/b5f863c67e6ffbaadedb94c885c3c50c625e4eb8/libcontainerd/remote/client.go#L465) -> `process.Delete`)

As far as I understand, the difference is that `Process.Delete` calls `p.client.TaskService().DeleteProcess()` *before* waiting for and closing `io` ([it happens here](https://github.com/containerd/containerd/blob/157fe6a/process.go#L225)), while `Task.Delete` will call `t.client.TaskService().Delete()` only *after* it does waiting and closing io ([here](https://github.com/containerd/containerd/blob/2aeac9a/task.go#L339)). My guess it that, after the deletion containerd shim exits and channels leading to dockerd are closed, so `io.Wait()` won't get stuck

I also noticed that `Process.Delete` was not always working like that, initially it also was calling `p.io.Wait()` *before* calling `DeleteProcess`. But this was changed by containerd commit [`fc535efe5845aa99f98b7fac58526b1b06c102ab`](https://github.com/containerd/containerd/commit/fc535efe5845aa99f98b7fac58526b1b06c102ab). It belongs to containerd/containerd#1394 - it seems that `Process.Delete` working that way was causing some tests to sometimes deadlock